### ISSUE TITLE
Module federation prevents to use expose-shared modules as singletons

### ISSUE THREAD

**Issue by smelukov** [webpack/webpack#18556]
**a.js**
```js
import {c} from './c'

export const a = c;
```

**b.js**
```js
import {c} from './c'

export const b = c;
```

**c.js**
```js
export const c = 123;
```

```js
new ModuleFederationPlugin({
  name: 'app2',
  exposes: {
    './App': './src/App',
    './a': './src/a',
    './b': './src/b',
  },
})
```

Here is an output after build
![image](https://github.com/webpack/webpack/assets/6654581/2631a376-17fe-4a11-b352-78ecdf967247)

`c.js` is using by `a` and `d` and was duplicated that is not expected behaviour
For example c.js may be some singleton-class but bundler-behaviour prevents it.
Is it a bug?

Repo https://github.com/smelukov/mf-duplicates/tree/main

**Comment by alexander-akait** [comment]
Yes, but you can move `c.js` forced into own chunk

> For example c.js may be some singleton-class but bundler-behaviour prevents it.

https://webpack.js.org/plugins/module-federation-plugin/#singleton

**Comment by smelukov** [comment]
A project may have many modules that should be used as singletons, and a developer never knows about all of them (e.g., packages in node_modules and their dependencies)
I think that the solution is to try to use split chunk logic for exposed modules used in a federated project itself
Current behaviour just breaks some projects and leads to unexpected errors

**Comment by alexander-akait** [comment]
@smelukov using another logic can break it too, so without manual setup it is impossible to solve

**Comment by ScriptedAlchemy** [comment]
Exposed modules dont know C needs to be anything special, you must set it as singleton otherwise it would bundle what that chunks needs like anything else.

**Comment by smelukov** [comment]
@ScriptedAlchemy 
A really huge project may have a tons of cross-expose imports and these modules should not be duplicated (components, data stores, etc)
Should I mark the entire project as a singleton or some directory?
For example:

```ts
new ModuleFederationPlugin({
  name: 'app2',
  exposes: {
    './components': './src/components',
    './pages': './src/pages',
  },
  shared: {
    './src/components/': { singleton: true }, // with trailing slash
    './src/pages/': { singleton: true }, // with trailing slash
  }
})
```

Will (should) it work that way?

**Comment by ScriptedAlchemy** [comment]
Share the thing they both need as a singleton. Sharing exposed modules as singletons won't work since an exposed module is a singleton, but the sub-dependencies in there are not, unless you mark them as such.

If you have a host where its own remote entry is consumed as well, then that's your actual issue. There are two module graphs on the page, and the instance in the host is a different runtime and module cache than the remote entry. In such a case, you need a single runtime chunk or can sometimes work around it by sharing and not exposing. But the reason it "works" is that the remote doesn't actually load a copy and the duplicate runtime happens to use one from the other. 

While the workaround works. If you do have host uses its own remote somehow in the tree. Then chunk loading will eventually break etc

**Comment by ScriptedAlchemy** [comment]
But if you use single runtime chunk you should upgrade to federation v2, since webpack currently has issue with single runtime and mf 

https://module-federation.io/

**Comment by smelukov** [comment]
@ScriptedAlchemy 
My case is:

**Host**
- has an entrypoint and one expose
- uses own expose (lets name it as hostStore)

**Remotes**
- use hostStore-expose from host
- contains some number of exposes and there is many cross-exposes imports in this remotes

I use a plugin from https://module-federation.io/

**Comment by ScriptedAlchemy** [comment]
Yeah then that's your problem. You have two runtimes from the same build. 
Set RuntimeChunk:single in optimization.

**Comment by smelukov** [comment]
1) changed import
from name+expose:
`import { Store } from 'host/hostStore';`
to relative:
`import { Store } from './hostStore';`

2) changed config:

```
optimization: {
    runtimeChunk: 'single',
},
```

Got error
![image](https://github.com/user-attachments/assets/e2a25fde-2c2b-4700-b370-c6166a2485ef)

**Comment by ScriptedAlchemy** [comment]
The problem is that you have two runtime chunks. We cannot auto-fix that. (It should be a) single runtime chunk for the host to load its own remote. Tomorrow I will see if I can ship a slight update to how it works. I have something to handle making both single runtime and standalone runtime. Then you use the runtime plugin to switch between which remote URL to use depending on if the remote is also the host or not.

Single runtime means the remote won't work off the host. Or without the host's parent runtime.
If your host has other dynamic imports that load after its remote, you will get chunk load timeout or failed to load errors from webpack. Since the remote will overwrite the hosts chunk loading global.

**Comment by ScriptedAlchemy** [comment]
Set experiments.federationRuntime = "hoisted"https://module-federation.io/configure/experiments.html

**Comment by smelukov** [comment]
@ScriptedAlchemy, do you have any plans or ideas for solving duplication cases? We have a really huge project with tons of MF bundles and exposes. If exposes use the same module, this module will be duplicated into two expose chunks.
For example, we have two forms with a form component. These forms use the button component. The button component will be duplicated in the final bundle.
